### DISCUSSION TITLE
Issue with a custom gateway plugin when Give version is 2.8+

### DISCUSSION THREAD

**Discussion by maximestd** [impress-org/givewp#6285] [Developer Q&A]
Hi, I developped a plugin to add a gateway to Give.

I have an issue when the donation form redirect to my gateway since the 2.8 version. When I click on the button to validate the donation, I am redirected to the url with "?payment-mode=[mygateway]" but doesn't redirect to the gateway url. 

In 2.7.5 version that works correctly, after the url with "?payment-mode=[mygateway]", I am redirected to my gateway url.

I tried to find what changes were done with 2.8 but without success.

Can a developer help me ?

More info : Currently I use the hook 'give_gateway_{$gateway}' to add a function who set up the data form which should be sent to the gateway and this data form has an "action= [the gateway redirect url]"

I am testing on a dev environment.

Thanks.

Edit : It looks like it's the "echo" of my form which doesn't work. I need to use POST method to send the form data so how can I do that ?

**Comment by jonwaldstein** [comment]
Hi @maximestd

We recognize how confusing and difficult creating a custom GiveWP gateway can be.  That why we've actually been working on a new gateway api that will make creating custom gateways much simpler.  It's technically included in GiveWP as of version 2.18.0/2.19.0 but we haven't announced it to the developer community yet because we're updating our own gateways to use it first.

That being said, your use case sounds like a good candidate for this new api.  Would you be interested in trying it out?  We could supply you with some documentation to get started but you might need to give us some more information or even code samples about what your building to make sure it's currently possible.

Let me know what you think,

Thanks

**Reply by maximestd** [reply]
Thanks for your answer, I'm interested in trying it.
For context, the plugins are developed by someone else. I took over the work of someone else who had taken it over from another to adapt it :')
At the start, the plugin was done to add a "Paybox" gateway to Give, but it was adapted to add a "La Banque Postale" gateway.
I resumed work because after the 2.8 update, the form is not sent.

This is some code that I use :

`add_action('give_gateway_paybox',              ['Give_Paybox_Gateway', 'process_payment']);
`
In process_payment(), after creating payment into Give we have :
```
$paybox_transaction = new Give_Paybox($payment_id);
$paybox_transaction::makeHMACForm($url, $params, $hmac);
```

This is how the form is created in makeHMACForm :

```
<html>
	<body>
		<form name="redirectForm" method="POST" action="<?= $url_prod_v3 ?>">
		 <input type="hidden" name="vads_action_mode" value="<?= $action_mode ?>" />
		 <input type="hidden" name="vads_amount" value="<?= $pbx_parameters ?>" />
		 <input type="hidden" name="vads_ctx_mode" value="<?= $ctx_mode ?>" />
		 <input type="hidden" name="vads_currency" value="<?= $currency ?>" />
		 <input type="hidden" name="vads_order_id" value="<?= $hmac ?>" />
		 <input type="hidden" name="vads_page_action" value="<?= $page_action ?>" />
		 <input type="hidden" name="vads_payment_cards" value="<?= $payment_cards ?>" />
		 <input type="hidden" name="vads_payment_config" value="<?= $payment_config ?>" />
		 <input type="hidden" name="vads_site_id" value="<?= $marchantv3 ?>" />
		 <input type="hidden" name="vads_trans_date" value="<?= $datetrans ?>" />
		 <input type="hidden" name="vads_trans_id" value="<?= $trans_id ?>" />
		 <input type="hidden" name="vads_version" value="<?= $vads_version ?>" />
		 <input type="hidden" name="signature" value="<?= $signature ?>" />
		 <noscript><input type="submit" name="Go" value="Click to continue" /></noscript>
	 	</form>
		<script type="text/javascript"> document.redirectForm.submit(); </script>
	</body>
</html>
```

If you want, I can maybe send you the plugins to understand better how that works.

Thanks

**Reply by jonwaldstein** [reply]
@maximestd okay got it.

Currently `add_action('give_gateway_$gateway')` will give you a big array of data that you would need to de-struct, insert the donation, and then make your calls to the gateway api.  It's pretty complicated.  I would need to look at your code to see if it's on the right track. If you want to give me access to your repo I can take a look and make some recommendations: `jon@givewp.com`

**Reply by maximestd** [reply]
@jonwaldstein 
Hi, I decided to start over by following this doc.
https://givewp.com/documentation/developers/how-to-build-a-gateway-add-on-for-givewp/

But I have the same problem. Here the code after inserting payment into Give.

In 2.7, the form is posted and redirect to my payment gateway url, in 2.8 and after, the payment is inserted in Give but the form is not sent.

` $donation_id = give_insert_payment( $donation_data );

		if ( ! $donation_id ) {

			// Record Gateway Error as Pending Donation in Give is not created.
			give_record_gateway_error(
				__( 'lbp Error', 'lbp-for-give' ),
				sprintf(
				/* translators: %s Exception error message. */
					__( 'Unable to create a pending donation with Give.', 'lbp-for-give' )
				)
			);

			// Send user back to checkout.
			give_send_back_to_checkout( '?payment-mode=lbp' );
			return;
		}

		// Do the actual payment processing using the custom payment gateway API. To access the GiveWP settings, use give_get_option()
                // as a reference, this pulls the API key entered above: give_get_option('insta_for_give_instamojo_api_key')

		//preparing data to the form 

							<html>
								<body>
									<form name="redirectForm" method="POST" action="<?= $url_prod_v3 ?>">
									 <input type="hidden" name="vads_action_mode" value="<?= $action_mode ?>" />
									 <input type="hidden" name="vads_amount" value="<?= $pbx_parameters ?>" />
									 <input type="hidden" name="vads_ctx_mode" value="<?= $ctx_mode ?>" />
									 <input type="hidden" name="vads_currency" value="<?= $currency ?>" />
									 <input type="hidden" name="vads_order_id" value="<?= $hmac ?>" />
									 <input type="hidden" name="vads_page_action" value="<?= $page_action ?>" />
									 <input type="hidden" name="vads_payment_cards" value="<?= $payment_cards ?>" />
									 <input type="hidden" name="vads_payment_config" value="<?= $payment_config ?>" />
									 <input type="hidden" name="vads_site_id" value="<?= $marchantv3 ?>" />
									 <input type="hidden" name="vads_trans_date" value="<?= $datetrans ?>" />
									 <input type="hidden" name="vads_trans_id" value="<?= $trans_id ?>" />
									 <input type="hidden" name="vads_version" value="<?= $vads_version ?>" />
									 <input type="hidden" name="signature" value="<?= $signature ?>" />
									 <noscript><input type="submit" name="Go" value="Click to continue" /></noscript>
								 	</form>
									<script type="text/javascript"> document.redirectForm.submit(); </script>
								</body>
							</html>
<?

	} else {

		// Send user back to checkout.
		give_send_back_to_checkout( '?payment-mode=lbp' );
	} // End if().
}

// change the insta_for_give prefix to avoid collisions with other functions.
add_action( 'give_gateway_lbp', 'lbp_for_give_process_lbp_donation' );`

Thanks.

**Reply by maximestd** [reply]
@jonwaldstein
Finally found the problem. in 2.8 version of Give, buffering was added, so my form was never sent.

`ob_start();
// Send info to the gateway for payment processing.
give_send_to_gateway( $donation_data['gateway'], $donation_data );
ob_get_clean();`

**Reply by jonwaldstein** [reply]
Thanks for the update.  Glad to hear its working.  We are actually in the process of writing documentation for our new gateway api that will simplify this whole process.

**Reply by emanuellopess** [reply]
@jonwaldstein Any news of the new documentation to build gateways for GiveWP 3.0? I'm got some progress but I need to know how to properly render inputs in the new form, the test gateway is utilizing a .tsx file

**Reply by jonwaldstein** [reply]
Hi @emanuellopess,

Yes, we have documentation on the new payment gateway api [here](https://givewp.com/documentation/developers/how-to-build-a-gateway-add-on-for-givewp/).  However, it is not complete with compatibility direction on building for GiveWP 3.0 yet.  I can give you some preliminary direction to get started:

In order to render fields in v3 forms, you need to use the `public function enqueueScript(int $formId)` [method](https://github.com/impress-org/givewp/blob/cb1412040f1e539020ec39338fe3c0c7b0ab6d78/src/PaymentGateways/Gateways/TestGateway/TestGateway.php#L51C21-L51C34) inside your Gateway class to load a javascript file using the WP function `wp_enqueue_script`.  Then inside that javascript file you would register your front-end gateway object:
```
window.givewp.gateways.register({
id: 'my-gateway-id',
Fields(){
    <input type="text" name="my-gateway-field-name" />
}
});
```
If you want to interact with your input fields, then you will want to use React.

A good place to start looking would be our Test Gateway:
- [TestGateway.php](https://github.com/impress-org/givewp/blob/develop/src/PaymentGateways/Gateways/TestGateway/TestGateway.php)
- [testGateway.tsx](https://github.com/impress-org/givewp/blob/develop/src/PaymentGateways/Gateways/TestGateway/testGateway.tsx)

For a real gateway, Stripe Payment Element is a good example:
- [StripePaymentElementGateway.php](https://github.com/impress-org/givewp/blob/develop/src/PaymentGateways/Gateways/Stripe/StripePaymentElementGateway/StripePaymentElementGateway.php)
- [stripePaymentElementGateway.tsx](https://github.com/impress-org/givewp/blob/develop/src/PaymentGateways/Gateways/Stripe/StripePaymentElementGateway/stripePaymentElementGateway.tsx)

Please let me know if any of this is making sense or is too confusing.  In the meantime, i'll get started on writing the documentation.

**Reply by emanuellopess** [reply]
@jonwaldstein Thanks for the fast response, how can I transpile the React files to work with the new GiveWP version? I saw it has some dependencies: 

```js
import { __ } from '@wordpress/i18n';
import type { FormData, Gateway } from '@givewp/forms/types';
```

Can I omit those? Or how may I include with those dependencies?

**Reply by jonwaldstein** [reply]
For React, we suggest using [@wordpress/scripts](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-scripts/) to transpile which uses react and the `@wordpress/` dependencies from the current WordPress installation (window). 

Unfortunately, you won't be able to use the typescript types inside GiveWP from `@givewp/` so you can omit those.
